merlin/core/dispatch.py

```diff
@@ -75,14 +76,6 @@ def inner2(*args, **kwargs):
         return inner1
 
 
-if HAS_GPU:
-    DataFrameType = Union[pd.DataFrame, cudf.DataFrame]  # type: ignore
-    SeriesType = Union[pd.Series, cudf.Series]  # type: ignore
-else:
-    DataFrameType = pd.DataFrame  # type: ignore
-    SeriesType = pd.Series  # type: ignore
-
```

Comment:
These types have been a pain since they don't play nicely with Mypy, which doesn't like dynamically defined types like this. This PR replaces them with [Python protocols](https://peps.python.org/pep-0544/) that can be used as a static types and also checked at runtime (with `isinstance`.)

Reply:
We could keep this code around for a follow-up PR to remove once the references to these in the other repos have been removed. That way we might be able to merge this without breaking tests in the other repos as a result, making the transition smoother?

Reply:
Yeah, that's a good idea 👍🏻